Overhaul documentation for declarative net request (#6921)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/reference/declarativeNetRequest/index.md

## Manifest {: #manifest }

### Permissions {: #permission-definitions }

`declarativeNetRequest`

> : Allows extensions to block and upgrade requests.

Reviewer: With how this is currently written, it implies to me that `declarativeNetRequest` is less capable than `declarativeNetRequestWithHostAccess`. Could we clarify that with sufficient host permissions both are equivalent (and can redirect requests and modify headers)?

Author: We can, but I need some clarification because this definition regurgitates the current definition. Your explanation makes it sound as though there's no difference between the permissions. 

What's the difference between the two permissions?

Reviewer: Hopefully our discussion yesterday was helpful :)

Reviewer: I took a look at the updated summary and I think it would still be nice to work on. It says "both require host permissions" which is not true - "declarativeNetRequestWithHostAccess" doesn't require anything, but for it to be useful you have to separately request host permissions for the sites you're interested in.

`declarativeNetRequestWithHostAccess`
: Allows extensions to redirect requests and modify request and response headers. This permission always requires [host permissions](/docs/extensions/mv3/declare_permissions/) to the requestED URL and initiator to act on a request.

`declarativeNetRequestFeedback`
: Allows access to functions and events that return information on matched rules during debugging. This permission only works in [unpacked extensions](/docs/extensions/mv3/getstarted/development-basics/#load-unpacked).

Certain rules such as redirect and header modification rules also require host permissions that match the network requests that these rules apply to.

> ### Static rulesets {: #manifest-rule-resources }

Reviewer: It feels like this should come after the concepts section where we explain what a ruleset is?

}
```

## Concepts and Usage {: #concepts-and-usage }

To use this API specify one or more rulesets. A ruleset contains an array of rules. A single rule does one of the following:

-  Block a network request.

> -  Upgrade a request.

Reviewer 2: Can you maybe add a short explanation what `upgrade` means (at least I don't know what upgrade means in this context).

Author: Done.

}
```

## Concepts and Usage {: #concepts-and-usage }

> To use declarative net requests specify one or more rulesets. A ruleset contains an array of rules. A single rule does one of the following:

Reviewer: I think "declarative net requests" should be either the API name or namespace, since it's not a thing that makes sense to pluralise.

> To use declarative net requests specify one or more rulesets. A ruleset contains an array of rules. A single rule does one of the following:

Reviewer: A rule can actually do five things (e.g upgrade requests). Should we say something that makes it clear this is not an exhaustive list?

-  Block a network request.
-  Prevent a request from getting blocked by negating any matching blocked rules.
-  Redirect a network request.
-  Modify request or response headers.

There are three types of rulesets.

> -  Static rules

Reviewer: ```suggestion
-  Static rulesets
```

etc.?

Reviewer: Hmm, now it says "There are three types of rulesets." and then lists "static rules". That feels wrong too?

Author: I thought I had removed the word 'rules'.

-  Dynamic rules
-  Session rules

### Dynamic and session-scoped rulesets {: #dynamic-and-session-rules }

Dynamic and session-scoped rulesets are managed using JavaScript while an extension is in use.

-  Dynamic rules persist across browser sessions and extension upgrades.

> -  Session rules are destroyed when the browser shuts down and when a new version of the extension is installed.

Reviewer: ```suggestion
-  Session rules are cleared when the browser shuts down and when a new version of the extension is installed.
```

#update-static-rulesets

#### Dynamic and session rules {: #dynamic-session-rules }

The limits applied to dynamic and session rules are simpler than static rules. The total number of both cannot exceed **5000**. This is called the [`MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES`][#property-MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES].

#### Rules that use regex {: #regex-rules }

All types of rules can use regular expressions; however the total number of regex rules, of each type cannot exceed **1000**. This is called the [MAX_NUMBER_OF_REGEX_RULES](#property-MAX_NUMBER_OF_REGEX_RULES).

### Interactions with service workers {: #interact-w-service-workers }

> Requests and responses intercepted by the `onfetch` handler in a host page's service worker are not affected by declarativeNetRequet. Consequently such service workers can make it appear as though a rule doesn't work. To properly test your rules, you may need to disable or remove one or more of the service workers in your installation of Chrome.

Reviewer 2: Do I understand this right that publishers could simply handle every request in the `onfetch` handler to avoid ad blockers functioning on their pages?

Reviewer: So I was trying to avoid opening this can of worms and I think the above text is still vague enough to sidestep this - but if you call fetch in a service worker, those requests are affected by DNR. So even if a web service worker intercepts a request, the only way it can meaningfully respond is (a) if it already has a response cached or (b) if it does a fetch to make the request anyway, and DNR will apply. Hopefully that makes sense.

Reviewer 2: We need to add this detail otherwise it is very misleading.

Author: That opens another can of worms. If I understand what you're saying correctly, whether and how dNR works depends on which [caching algorithm](https://web.dev/offline-cookbook/#serving-suggestions) a host page uses.

- *Cache only*- dNR will always fail
- *Network only*- dNR will always succeed
- *Cache, falling back to network*- dNR will fail when the item is in the cache, succeed otherwise
- *Cache and network race*- Who knows what dNR will do?
- *Network falling back to cache*- Whenever there's an item in the cache, dNR will appear to fail, but succeed otherwise.
- *Generic fallback*- dNR will appear to fail when the network can't be reached.
- *Service worker templating*- dNR will appear to fail.

I can't see how an extension developer would design around this.

Author: There must be some very good reason why dNR doesn't affect items retrieved from `CacheStorage`.

Reviewer: @jpmedley - all of the above sounds right 👍 

I don't feel quite as pessimistic about being able to document this. How about the following (changes in bold)?

> Requests and responses intercepted by the `onfetch` handler in a host page's service worker are not affected by declarativeNetRequest. Consequently such service workers can make it appear as though a rule doesn't work. To properly test your rules, you may need to disable or remove one or more of the service workers in your installation of Chrome. **Note that DNR does apply if the service worker goes on to make subsequent requests to the network, even if these are to the same path that was previously intercepted.**

On your other point:

> I can't see how an extension developer would design around this.

I think this depends on the developer's goal. If you're making something to block trackers, a request to /track_user that never reaches the network is useless, so the developers of those websites aren't going to enable any caching. In that case you can probably just ignore all of this.

If you are trying to make something to block a page from loading, you have bigger problems than this. DNR doesn't have any affect on single page apps that use `history.pushState` to change what is showing in the address bar for example.

So TLDR, I think making developers aware of this gotcha is important, but I don't expect it actually causes many issues in practice.

Reviewer 2: I don't thinks this phrasing is correct on it's own:

> Requests and responses intercepted by the `onfetch` handler in a host page's service worker are not affected by declarativeNetRequest. 

Because they are affected, if the sw makes a subsequent fetch request. I still like Oliver's original phrasing, we could add an example to make it more clear:

```
DNR only applies to requests that reach the network stack. This does include responses from the HTTP cache, but does not include responses generated by a website's service worker. For example, if a service worker intercepts a request in an `onFetch` handler and returns a cached or generated response, the response will not be affected by your DNR rules. However, if the service worker makes a subsequent `fetch` request, your DNR rules will apply to this new request.
```

> declarativeNetRequest only applies to requests that reach the network stack. This includes responses from the HTTP cache, but may not include responses that go through a service worker's `onfetch` handler. declarativeNetRequest won't affect responses generated by the service worker or retrieved from `CacheStorage`, but it will affect calls to `fetch()` made in a service worker.

Author: New language.

@oliverdunk @sebastianbenz

Reviewer 2: Looks good!

Reviewer: Sounds great! 🎉

### Rule limits {: #limits }

There is a performance overhead to loading and evaluating rules in the browser,
and so a number of limits apply when using the API. Limits depend on the type of
rule you're using.

#### Static rules {: #static-rules }

Static rules are those specified in rule files (which are specified in the manifest file).

An extension can specify up to **50** static [rulesets](#type-Ruleset) as part of the `"rule_resources"` manifest key, but only **10** of these rulesets can be enabled at a time. The latter is called the [`MAX_NUMBER_OF_ENABLED_STATIC_RULESETS`](#property-MAX_NUMBER_OF_ENABLED_STATIC_RULESETS). Collectively, those rulesets are guaranteed at least **30,000** rules. This is called the [`GUARANTEED_MINIMUM_STATIC_RULES`](#property-GUARANTEED_MINIMUM_STATIC_RULES). The number of rules available after that depends on how many rules are enabled by all the extensions installed on a user's browser. You can find this number at runtime by calling [`getAvailableStaticRuleCount()`](#method-getAvailableStaticRuleCount). You can see [an example of this](#update-static-rulesets) under [code examples](#code-examples).

> #update-static-rulesets

Reviewer 2: Is this supposed to be a heading?

Author: Removed. It's an old anchor.

#### Dynamic and session rules {: #dynamic-session-rules }

> The limits applied to dynamic and session rules are simpler than static rules. The total number of both cannot exceed **5000**. This is called the [`MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES`][#property-MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES].

Reviewer 2: ```suggestion
The limits applied to dynamic and session rules are simpler than static rules. The total number of both cannot exceed **5000**. This is called the [`MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES`](#property-MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES).
```

Author: Fixed.

#### Rules that use regex {: #regex-rules }

All types of rules can use regular expressions; however the total number of regex rules, of each type cannot exceed **1000**. This is called the [MAX_NUMBER_OF_REGEX_RULES](#property-MAX_NUMBER_OF_REGEX_RULES).

Additionally, each rule must be less than 2KB once compiled. This roughly correlates with the complexity of the rule. If you try to load a rule that exceeds this limit, you will see a warning like the one below and the rule will be ignored.

```bash

> rules_1.json: Rule with id 1 specified a more complext regex than allowed as part of the "regexFilter" key.

Reviewer: ```suggestion
rules_1.json: Rule with id 1 specified a more complex regex than allowed as part of the "regexFilter" key.
```

> rules_1.json: Rule with id 1 specified a more complext regex than allowed as part of the "regexFilter" key.

Reviewer 2: Can we make the `pre` wrap the content here? Otherwise maybe add linebreaks.

Author: Done.

  let options = { enableRulesetIds: enableRulesetIds }
  // Get the number of enabled static rules
  const enabledStaticCount = await chrome.declarativeNetRequest.getEnabledRulesets();
  // Compare rule counts to determine if anything needs to be disabled so that
  // new rules can be enabled
  const propsedCount = enableRulesetIds.length;
  if (enabledStaticCount + proposedCount > chrome.declarativeNetRequest.MAX_NUMBER_OF_ENABLED_STATIC_RULESETS) {
    options.disableRulesetIds = disableCandidateIds
  }
  // Update the enabled static rules
  await chrome.declarativeNetRequest.updateEnabledRulesets(options);
}
```

### Rule examples {: #rule-examples }
The examples below illustrate how Chrome prioritizes rules in an extension. You may want to open the [prioritization](#implementation-matching-algorithm) rules in a separate window when reviewing them.

#### The "priority" key {: #priority-key }

These examples require the following host permission:

`*://*.example.com/*`

To work out the priority of a particular URL, look at the (developer-defined) `"priority"` key, the `"action"` key and the `"urlFilter"` key. These examples refer to the example rule file shown below them.

> **Navigation to https://google.com**

Reviewer 2: The formatting looks a bit weird when rendered to html as the linebreak disappears.

Author: Fixed.

```js
async function updateStaticRules(enableRulesetIds, disableCandidateIds) {
  // Create the options structure for the call to updateEnabledRulesets()
  let options = { enableRulesetIds: enableRulesetIds }
  // Get the number of enabled static rules
  const enabledStaticCount = await chrome.declarativeNetRequest.getEnabledRulesets();
  // Compare rule counts to determine if anything needs to be disabled so that
  // new rules can be enabled
  const proposedCount = enableRulesetIds.length;
  if (enabledStaticCount + proposedCount > chrome.declarativeNetRequest.MAX_NUMBER_OF_ENABLED_STATIC_RULESETS) {
    options.disableRulesetIds = disableCandidateIds
  }
  // Update the enabled static rules
  await chrome.declarativeNetRequest.updateEnabledRulesets(options);
}
```

### Rule examples {: #rule-examples }
The examples below illustrate how Chrome prioritizes rules in an extension. When reviewing them, you may want to open the [prioritization](#implementation-matching-algorithm) rules in a separate window.

#### The "priority" key {: #priority-key }

> These examples require [host permission](/docs/extensions/mv3/declare_permissions/) to `*://*.example.com/*`.

Reviewer 3: > These examples require [host permission](/docs/extensions/mv3/declare_permissions/) to `*://*.example.com/*`.

@jpmedley Are we sure the rules below don't need to declare "https://*.google.com/*" as a host permission.

Author: They do.

> These examples require [host permission](/docs/extensions/mv3/declare_permissions/) to `*://*.example.com/*`.

Author: ```suggestion
These examples require [host permission](/docs/extensions/mv3/declare_permissions/) to `https://.google.com/`.
```

    </tbody>
</table>

### Rule prioritization {: #implementation-matching-algorithm }

Rules are triggered by requests sent from web pages. If multiple rules match a particular request, then the rules must be prioritized. This section explains how they're prioritized. Prioritization happens in two stages.

1. Priority is determined for rules within an extension.
1. If more than one extension can apply a rule to a request, priority is determined for all extensions that match a particular request.

> Thinking of matching this way: whatever rule a particular extension prioritizes will then be prioritized against rules from other extensions.

Author: The result of the first part is a single rule. The word 'the' is not needed to make this grammatically correct.

#### Rules that use regex {: #regex-rules }

All types of rules can use regular expressions; however, the total number of regex rules of each type cannot exceed 1000. This is called the [MAX_NUMBER_OF_REGEX_RULES](#property-MAX_NUMBER_OF_REGEX_RULES).

Additionally, each rule must be less than 2KB once compiled. This roughly correlates with the complexity of the rule. If you try to load a rule that exceeds this limit, you will see a warning like the one below and the rule will be ignored.

```bash
rules_1.json: Rule with id 1 specified a more complext regex than allowed
as part of the "regexFilter" key.
```

### Interactions with service workers {: #interact-w-service-workers }

declarativeNetRequest only applies to requests that reach the network stack. This includes responses from the HTTP cache, but may not include responses that go through a service worker's `onfetch` handler. declarativeNetRequest won't affect responses generated by the service worker or retrieved from `CacheStorage`, but it will affect calls to `fetch()` made in a service worker.

### Web accessible resources {: #implementation-web-accessible-resources }

> A declarativeNetRequest rule cannot redirect from a public resource request to a resource that is not web accessible. Doing so triggers an error. This is true even if the specified web accessible resource is owned by the redirecting extension. To declare resources for declarativeNetRequests, use the manifest's [`"web_accessible_resources"`](/docs/extensions/mv3/manifest/web_accessible_resources/) array.

Reviewer: ```suggestion
A declarativeNetRequest rule cannot redirect from a public resource request to a resource that is not web accessible. Doing so triggers an error. This is true even if the specified web accessible resource is owned by the redirecting extension. To declare resources for declarativeNetRequest, use the manifest's [`"web_accessible_resources"`](/docs/extensions/mv3/manifest/web_accessible_resources/) array.
```

const oldRuleIds = oldRules.map(rule => rule.id);

// Use the arrays to update the dynamic rules
await chrome.declarativeNetRequest.updateDynamicRules({
  removeRuleIds: oldRuleIds,
  addRules: newRules
});
```

#### Update static rulesets {: #update-static-rulesets }

The following example shows how to enable and disable rulesets, while taking into account the number of available rulesets and the maximum number of enabled static rulesets. You would do this is in situations where the number of static rules you need exceeds the number allowed. For this to work, some of your rulesets should be installed with some of your rulesets disabled (setting `"Enabled"` to `false` with in the manifest file).

```js
function updateStaticRules(enableRulesetIds, disableCandidateIds) {
  // Create the options structure for the call to updateEnabledRulesets()
  let options = { enableRulesetIds: enableRulesetIds }
  // Get the number of enabled static rules
  const enabledStaticCount = await chrome.declarativeNetRequest.getEnabledRulesets();
  // Compare rule counts to determine if anything needs to be disabled so that
  // new rules can be enabled

>   const propsedCount = enableRulesetIds.length;

Reviewer 3: ```suggestion
  const proposedCount = enableRulesetIds.length;
```

  if (enabledStaticCount + proposedCount > chrome.declarativeNetRequest.MAX_NUMBER_OF_ENABLED_STATIC_RULESETS) {
    options.disableRulesetIds = disableCandidateIds
  }
  // Update the enabled static rules
  await chrome.declarativeNetRequest.updateEnabledRulesets(options);
}
```

### Rule examples {: #rule-examples }
The examples below illustrate how Chrome prioritizes rules in an extension. You may want to open the [prioritization](#implementation-matching-algorithm) rules in a separate window when reviewing them.

#### The "priority" key {: #priority-key }

These examples require the following host permission:

> `*://*.example.com/*`

Reviewer 3: > `*://*.example.com/*`

Is this the only permission required for the example below? It might be nice to have a full manifest code sample for ease of copy/paste

To work out the priority of a particular URL, look at the (developer-defined) `"priority"` key, the `"action"` key and the `"urlFilter"` key. These examples refer to the example rule file shown below them.

Navigation to https://google.com
: Two rules cover this URL: the rules with IDs 1 and 4. The rule with ID 1 applies because `"block"` actions have a higher priority than `"redirect"` actions. The remaining rules do not apply because they are for longer URLs.

Navigation to https://google.com/1234

> : Because of the longer URL, the rule with ID 2 now matches in addition to  the rules with IDs 1 and 4. The rule with ID 2 applies because `"allow"` has a higher priority than `"block"` and `"redirect"`.

Reviewer 3: ```suggestion
: Because of the longer URL, the rule with ID 2 now matches in addition to the rules with IDs 1 and 4. The rule with ID 2 applies because `"allow"` has a higher priority than `"block"` and `"redirect"`.
```

    </tbody>
</table>

### Rule prioritization {: #implementation-matching-algorithm }

Rules are triggered by requests sent from web pages. If multiple rules match a particular request, then the rules must be prioritized. This section explains how they're prioritized. Prioritization happens in two stages.

1. Priority is determined for rules within an extension.
1. If more than one extension can apply a rule to a request, priority is determined for all extensions that match a particular request.

Thinking of matching this way: whatever rule a particular extension prioritizes will then be prioritized against rules from other extensions.

{% Aside %}

> Avoid building rulesets that depend on rules with the same priority running in a particular order. Rules with the same priority in the same set (e.g static rules with priority 1) are executed in an arbitrary order which is not defined and can change between runs or browser versions. This guidline applies between browsers as well.

Reviewer: ```suggestion
Avoid building rulesets that depend on rules with the same priority running in a particular order. Rules with the same priority in the same set (e.g static rules with priority 1) are executed in an arbitrary order which is not defined and can change between runs or browser versions. This is the case across browsers.
```

{% endAside %}

#### Rule prioritization within an extension {: #rule-prioritization-within-an-extension }

Within a single extension, prioritization is worked out using the following process:

1. The rule with the highest developer-defined priority (in other words, the `"priority"` field) is returned.
1. If there is more than one rule with the highest developer-defined priority, rules are prioritized using the `"action"` field, in the following order:

>     -  `allow`

Reviewer: Maybe an numbered list would make sense here?

    -  `allowAllRequests`
    -  `block`
    -  `upgradeScheme`
    -  `redirect`

> 1. If the action type is not `block` or `redirect`, any matching `modifyHeaders` rules are evaluated. Be aware that if there are any rules with a developer-defined priority lower than the priority specified for `allow` and `allowAllRequests`, such rules are ignored.

Reviewer: TIL. Are we comfortable this is true?

Reviewer: Actually, I dug in to this more and it looks like you're right. TIL!

An extension can specify up to **50** static [rulesets](#type-Ruleset) as part of the `"rule_resources"` manifest key, but only **10** of these rulesets can be enabled at a time. The latter is called the [`MAX_NUMBER_OF_ENABLED_STATIC_RULESETS`](#property-MAX_NUMBER_OF_ENABLED_STATIC_RULESETS). Collectively, those rulesets are guaranteed at least **30,000** rules. This is called the [`GUARANTEED_MINIMUM_STATIC_RULES`](#property-GUARANTEED_MINIMUM_STATIC_RULES). The number of rules available after that depends on how many rules are enabled by all the extensions installed on a user's browser. You can find this number at runtime by calling [`getAvailableStaticRuleCount()`](#method-getAvailableStaticRuleCount). You can see [an example of this](#update-static-rulesets) under [code examples](#code-examples).

#update-static-rulesets

#### Dynamic and session rules {: #dynamic-session-rules }

The limits applied to dynamic and session rules are simpler than static rules. The total number of both cannot exceed **5000**. This is called the [`MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES`][#property-MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES].

#### Rules that use regex {: #regex-rules }

> All types of rules can use regular expressions; however the total number of regex rules, of each type cannot exceed **1000**. This is called the [MAX_NUMBER_OF_REGEX_RULES](#property-MAX_NUMBER_OF_REGEX_RULES).

Reviewer: There's also a memory limit on RegEx rules (I believe 2KB) that we should check and document.

Reviewer: Confirmed that the memory is limited to 2KB here: https://source.chromium.org/chromium/chromium/src/+/refs/heads/main:extensions/browser/api/declarative_net_request/utils.cc;l=317;drc=0340dd86c6c1f01ccd1a0b983633cfef5fcf6697

You get an error that looks like this (which is somewhat helpful, although it doesn't tell you what "complex" means behind the scenes 😔):

<img width="525" alt="Screenshot 2023-08-01 at 11 03 21" src="https://github.com/GoogleChrome/developer.chrome.com/assets/6097064/d4d7ec01-2f2b-4a10-89ae-8294f3ddf0c0">

Author: I'm open to wording suggestions.

Reviewer: > Additionally, each rule must be less than 2KB once compiled. This correlates with the general complexity of the rule specified. If you try to load a rule which exceeds this limit, you will see a warning and the rule will be ignored.

First draft to work from :)

Author: Added the paragraph, plus an example of the warning (in case someone google's the text of the warning).

Reviewer: Love it!

### Interaction with cached pages {: #implementation-cached-pages }

> If a rule matches a resource stored in cache by a host page's web service worker, that rule may be ignored until the resource is cleared from the cache. This is because cached storage is intended to be persistent, and many features such as offline use do not expect the cache to be cleared without also clearing a service worker's registration. For your rules to work properly, you need to call [`browsingData.removeCache()`](/docs/extensions/reference/browsingData/#method-removeCache) or [`browsingData.removeCacheStorage()`](/docs/extensions/reference/browsingData/#method-removeCacheStorage) after installation of your extension. You may need to call one of these functions after an extension upgrade, depending on whether you've changed your rules.

Reviewer: Note: I think we still need to discuss this one.

### Web accessible resources {: #implementation-web-accessible-resources }

> A declarative net request cannot redirect from a public resource request to a resource that is not web accessible. Doing so triggers an error. This is true even if the specified web accessible resource is owned by the redirecting extension. To declare resources for declarative net requests, use the manifest's [`"web_accessible_resources"`](/docs/extensions/mv3/manifest/web_accessible_resources/) array.

Reviewer: ```suggestion
A declarative net request rule cannot redirect from a public resource request to a resource that is not web accessible. Doing so triggers an error. This is true even if the specified web accessible resource is owned by the redirecting extension. To declare resources for declarative net requests, use the manifest's [`"web_accessible_resources"`](/docs/extensions/mv3/manifest/web_accessible_resources/) array.
```

To work out the priority of a particular URL, look at the (developer-defined) `"priority"` key, the `"action"` key and the `"urlFilter"` key. These examples refer to the example rule file shown below them.

**Navigation to https://google.com**

> : Two rules cover this URL: the rules with IDs 1 and 4. The rule with ID 1 applies because `"block"` actions have a higher priority than `"redirect"` actions. The remaining rules do not apply because they are for longer URLs.

Reviewer 2: There are now two subsequent colons in this paragraph

Author: This is how that renders:
![image](https://github.com/GoogleChrome/developer.chrome.com/assets/8204171/e16a1fe9-5c05-41ad-80e6-0f15d9efb0af)

Reviewer 2: an nice, all good then

    "action": { "type": "redirect", "redirect": { "url": "https://example.com" } },
    "condition": { "urlFilter": "google.com", "resourceTypes": ["main_frame"] }
  },
]
```

#### Redirects {: #redirects }

To use the example file shown below, your manifest file would need the following host permission:

`*://*.example.com/*`

> The following example shows how to redirect a request from example.com to a page within the extension itself. The extension path `/a.jpg` resolves to `chrome-extension://EXTENSION_ID/a.jpg`, where `EXTENSION_ID` is the ID of your extension.

Reviewer 3: It might be a good idea to mention that the a.jpg also needs to be declared as a WAR. Just in case a developers skipped the WAR section.

Author: Done.

```json
{
  "id": 1,
  "priority": 1,
  "action": { "type": "redirect", "redirect": { "extensionPath": "/a.jpg" } },
  "condition": {
    "urlFilter": "https://www.example.com",
    "resourceTypes": ["main_frame"]
  }
}
```

The following uses the `"transform"` key to redirect to a subdomain of example.com. It uses a domain name anchor ("||") to intercept requests with any scheme from example.com. The `"scheme"` key in `"transform"` specifies that redirects to the subdomain will always use "https".

```json
{
  "id": 1,
  "priority": 1,
  "action": {
    "type": "redirect",
    "redirect": {
      "transform": { "scheme": "https", "host": "new.example.com" }
    }
  },
  "condition": {
    "urlFilter": "||example.com",
    "resourceTypes": ["main_frame"]
  }
}
```

> The following example uses regular expressions to redirect from `https://www.abc.xyz.com/path` to `https://abc.xyz.com/path`. In the `"regexFilter"` key notice how periods are escaped and that the capturing group selects either "abc" or "def". The `"regexSubstitution"` key specifies the first returned match of the regular expression using "\\1". In this case "abc" is captured from the redirected URL and placed in the substitution.

Reviewer 3: ```suggestion
The following example uses regular expressions to redirect from `https://www.abc.xyz.com/path` to `https://abc.xyz.com/path`. In the `"regexFilter"` key, notice how periods are escaped and that the capturing group selects either "abc" or "def". The `"regexSubstitution"` key specifies the first returned match of the regular expression using "\\1". In this case, "abc" is captured from the redirected URL and placed in the substitution.
```

```json
{
  "id": 1,
  "priority": 1,
  "action": {
    "type": "redirect",
    "redirect": {
      "regexSubstitution": "https://\\1.xyz.com/"
    }
  },
  "condition": {
    "regexFilter": "^https://www\\.(abc|def)\\.xyz\\.com/",
    "resourceTypes": [
      "main_frame"
    ]
  }
}
```

#### Headers {: #headers }

> The following example removes all cookies from both a main frame and any sub frames.

Reviewer 3: ```suggestion
The following example removes all cookies from a main frame and sub frames.
```

Author: What's your reasoning? Mine sounds more specific to my ears.
